library/std/src/sys/unix/futex.rs

```diff
@@ -21,15 +21,16 @@ pub fn futex_wait(futex: &AtomicI32, expected: i32, timeout: Option<Duration>) {
             tv_nsec: d.subsec_nanos() as _,
         })
     });
-    unsafe {
+    let r = unsafe {
         libc::syscall(
             libc::SYS_futex,
             futex as *const AtomicI32,
             libc::FUTEX_WAIT | libc::FUTEX_PRIVATE_FLAG,
```

Comment:
`futex_wait` can cause shorter-than-expected timeouts if a signal occurs. A more robust approach would be to calculate the timeout as an absolute `Instant` and then use `FUTEX_WAIT_BITSET` which takes an absolute timeout. The wait itself should be looped so that it is restarted if it returns `EINTR`.

From the futex manpage:
> Note: for FUTEX_WAIT, timeout is interpreted as a relative value.  This differs from other futex operations, where timeout is interpreted as an absolute value.  To obtain the equivalent of FUTEX_WAIT with an absolute timeout, employ FUTEX_WAIT_BITSET with val3 specified as FUTEX_BITSET_MATCH_ANY.

Reply:
Does that matter much? According to the [signal (7)](https://man7.org/linux/man-pages/man7/signal.7.html) man page, FUTEX_WAIT is automatically restarted with SA_RESTART.

Reply:
There are 2 issues with that:
- The futex syscall will be restarted with the remaining time at the point where it was interrupted by a signal. This will be inaccurate if the signal handler takes a significant amount of time.
- The signal handler may not have been registered with `SA_RESTART`, in which case we would return too early.

Reply:
Updated in https://github.com/rust-lang/rust/pull/95035/commits/87299298d925af0943c94d2cc5bb8a2711d9f6b4